src/LaunchDarkly.EventSource/EventSource.cs

```diff
@@ -138,28 +144,67 @@ public async Task StartAsync()
                     }
                     await MaybeWaitWithBackOff();
                 }
-                firstTime = false;
+                firstTime = false;
+
+                CancellationTokenSource newRequestTokenSource = null;
+                CancellationToken cancellationToken;
+                lock (this)
+                {
+                    if (_readyState == ReadyState.Shutdown)
+                    {
+                        // in case Close() was called in between the previous ReadyState check and the creation of the new token
+                        return;
+                    }
+                    newRequestTokenSource = new CancellationTokenSource();
+                    _currentRequestToken?.Dispose();
+                    _currentRequestToken = newRequestTokenSource;
+                }
+
+                if (ReadyState == ReadyState.Connecting || ReadyState == ReadyState.Open)
+                {
+                    throw new InvalidOperationException(string.Format(Resources.EventSource_Already_Started, ReadyState));
+                }
+
+                SetReadyState(ReadyState.Connecting);
+                cancellationToken = newRequestTokenSource.Token;
+
                 try
                 {
-                    CancellationTokenSource newRequestTokenSource = null;
-                    lock (this)
-                    {
-                        if (_readyState == ReadyState.Shutdown)
-                        {
-                            // in case Close() was called in between the previous ReadyState check and the creation of the new token
-                            return;
-                        }
-                        newRequestTokenSource = new CancellationTokenSource();
-                        _currentRequestToken?.Dispose();
-                        _currentRequestToken = newRequestTokenSource;
-                    }
-                    await ConnectToEventSourceAsync(newRequestTokenSource.Token);
+                    await ConnectToEventSourceAsync(cancellationToken);
+
+                    // ConnectToEventSourceAsync normally doesn't return, unless it detects that the request has been cancelled.
+                    Close(ReadyState.Closed);
                 }
                 catch (Exception e)
-                {
-                    _logger.ErrorFormat("Encountered an error connecting to EventSource: {0}", e, e.Message);
-                    _logger.Debug("", e);
-                }
```

Comment:
I removed this ERROR-level logging because we are already giving the caller a way to be notified of errors, so they can easily decide whether they want to make those errors so prominent in their logs or not. ERROR level can have specific consequences for system monitoring, so in a very general-purpose library like this I think it's best not to make assumptions about how they want such things reported. In the context of the LaunchDarkly SDK, we are already doing our own error logging there in response to the Error event, so we were getting everything twice.